File: WoofIOS/SwiftPackages/Domain/UseCase/Sources/UseCase/Woofs/SendWoofUseCase.swift
import Combine
import DI
import Models
import Repositories

@Factory
public final class SendWoofUseCase {
    private let woofsRepository: WoofsRepository

    public func callAsFunction(dog: Dog) -> AnyPublisher<Void, WoofError> {
        woofsRepository.woofedDogIds
            .first()
            .setFailureType(to: WoofError.self)
            .flatMap { [woofsRepository] woofedIds -> AnyPublisher<Void, WoofError> in
                switch woofedIds.contains(dog.id) {
                case true: Fail(error: WoofError.alreadyWoofed).eraseToAnyPublisher()
                case false: woofsRepository.addWoof(dog: dog)
                }
            }
            .eraseToAnyPublisher()
    }
}

File: WoofAndroid/domain/usecase/src/main/kotlin/com/perrystreet/woof/usecase/woofs/SendWoofUseCase.kt
package com.perrystreet.woof.usecase.woofs

import com.perrystreet.woof.models.dog.Dog
import com.perrystreet.woof.models.errors.WoofException
import com.perrystreet.woof.repositories.woofs.WoofsRepository
import io.reactivex.rxjava3.core.Completable
import org.koin.core.annotation.Factory

@Factory
class SendWoofUseCase(
    private val woofsRepository: WoofsRepository,
) {
    operator fun invoke(dog: Dog): Completable =
        woofsRepository.woofedDogIds
            .firstOrError()
            .flatMapCompletable { woofedIds ->
                when (dog.id in woofedIds) {
                    true -> Completable.error(WoofException.AlreadyWoofed)
                    false -> woofsRepository.addWoof(dog)
                }
            }
}
